Added problem and solution for usaco-1188 (#3588)
cpinitiative/usaco-guide: solutions/platinum/usaco-1188.mdx

---
id: usaco-1188
source: USACO Platinum 2022 January
title: Minimizing Haybales
author: Aakash Gokhale
---

> [External Solution](http://www.usaco.org/current/data/sol_prob1_platinum_jan22.html)

Reviewer: ```suggestion
[External Solution(C++)](http://www.usaco.org/current/data/sol_prob1_platinum_jan22.html)
```

> ## Solution

Reviewer: ```suggestion
## Explanation
```

Notice that there is an edge case here where $a[i]$ can never be swapped down to the left past another element that is greater than it because of the restriction of $k$ even thought the $MAX$ element is greater than $a[i]$. In this case, we can push $a[i]$ into the next block and then apply the strategy of checking when the $MAX > a[i]$. We can check this by seeing that if after inserting $a[i]$ into the block and swapping towards the left and right appropriately as described above, $a[i]$ reaches the end of the block.

Now the problem arises that when we add enough elements, the size of a block could potentially be bigger than $\sqrt{n}$. To deal with this issue we can simply split the block into two smaller blocks (each half the size of the original block) and insert them into the list of blocks. Inserting into a vector in C++ or ArrayList in Java, is dependent on the size of the array which in our case which takes $\mathcal{O}(\sqrt{N})$. We can prove that the amount of blocks is bounded to $2\sqrt{n}$.

Processing each element of the array took $\mathcal{O}(\sqrt{N})$ time so the total time complexity is $\mathcal{O}(N\sqrt{N})$.

> 

Reviewer: ```suggestion

## Implementation

```

**Time Complexity:** $\mathcal{O}(N\sqrt{N})$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int n, k;

struct Block {
	int hi, lo;
	vector<int> vals;

> 	Block() {
> 		hi = -1e9;
> 		lo = 1e9;
> 	}

Reviewer: ```suggestion
	Block() : hi(-INT32_MAX), lo(INT32_MAX) { }
```

	Block(vector<int> a) {
		hi = -1e9;

> 		lo = 1e9;

Reviewer: why not just use the builtin constants here asw

> 		for (auto &i : a) {

Reviewer: ```suggestion
		for (int i : a) {
```clearer tbh

			vals.push_back(i);
			hi = max(hi, i);
			lo = min(lo, i);
		}
	}

> 	void add(int x) {

Reviewer: newlines between stuff

		lo = min(lo, x);
		hi = max(hi, x);
		vals.push_back(x);
		int i = vals.size() - 1;
		while (i > 0 && abs(vals[i] - vals[i - 1]) <= k) {
			swap(vals[i], vals[i - 1]);
			i--;
		}
		while (i < vals.size() - 1 && vals[i] > vals[i + 1]) {
			swap(vals[i], vals[i + 1]);
			i++;

> 		}

Reviewer: comments here

	}

> 	bool reaches_end(int x) {

Reviewer: wat dis function do (docstring)

		vals.push_back(x);
		int i = vals.size() - 1;
		while (i > 0 && abs(vals[i] - vals[i - 1]) <= k) {
			swap(vals[i], vals[i - 1]);
			i--;
		}
		while (i < vals.size() - 1 && vals[i] > vals[i + 1]) {
			swap(vals[i], vals[i + 1]);
			i++;
		}
		bool ans = vals.back() == x;
		vals.erase(begin(vals) + i);
		return ans;
	}
};

int main() {
	ios::sync_with_stdio(0);

> 	cin.tie(0);

Reviewer: remove if runs w/o

	cin >> n >> k;
	int len = (int)floor(sqrt(n));
	vector<Block> sqrt{Block()};
	for (int i = 0; i < n; i++) {
		int x;
		cin >> x;
		int j = sqrt.size() - 1;
		while (j > 0 && (x + k >= sqrt[j].hi && x - k <= sqrt[j].lo)) { j--; }
		if (j < sqrt.size() - 1 && sqrt[j].reaches_end(x)) {
			j++;
			while (j < sqrt.size() - 1 && x >= sqrt[j].hi) { j++; }
		}

> 		sqrt[j].add(x);

Reviewer: newlines between relevant code segments

		if (sqrt[j].vals.size() > len) {
			int half = (sqrt[j].vals.size()) / 2;
			vector<int> left, right;
			for (int k = 0; k < half; k++) { left.push_back(sqrt[j].vals[k]); }
			for (int k = half; k < sqrt[j].vals.size(); k++) {
				right.push_back(sqrt[j].vals[k]);
			}
			sqrt.insert(begin(sqrt) + j, Block(left));
			sqrt[j + 1] = Block(right);
		}
	}
	for (auto block : sqrt) {

> 		for (auto num : block.vals) { cout << num << endl; }

Reviewer: plz no auto

Now the problem arises that when we add enough elements, the size of a block could potentially be bigger than $\sqrt{n}$. To deal with this issue we can simply split the block into two smaller blocks (each half the size of the original block) and insert them into the list of blocks. Inserting into a vector in C++ or ArrayList in Java, is dependent on the size of the array which in our case which takes $\mathcal{O}(\sqrt{N})$. We can prove that the amount of blocks is bounded to $2\sqrt{n}$.

Processing each element of the array took $\mathcal{O}(\sqrt{N})$ time so the total time complexity is $\mathcal{O}(N\sqrt{N})$.

## Implementation

**Time Complexity:** $\mathcal{O}(N\sqrt{N})$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int n, k;

struct Block {

> 	int hi, lo;
> 	vector<int> vals;
> 	Block() : hi(-INT32_MAX), lo(INT32_MAX) {}
> 

Reviewer: ```suggestion
	int hi = -INT32_MAX;
	int lo = INT32_MAX;
	vector<int> vals;

```

	Block(vector<int> a) {
		hi = -INT32_MAX;
		lo = INT32_MAX;
		for (int i : a) {
			vals.push_back(i);
			hi = max(hi, i);
			lo = min(lo, i);
		}
	}

> 	// Adds element x to the block

Reviewer: ```suggestion
	/** Adds element x to the block */
```

	void add(int x) {
		lo = min(lo, x);
		hi = max(hi, x);
		vals.push_back(x);
		int i = vals.size() - 1;
		while (i > 0 && abs(vals[i] - vals[i - 1]) <= k) {
			swap(vals[i], vals[i - 1]);
			i--;
		}
		while (i < vals.size() - 1 && vals[i] > vals[i + 1]) {
			swap(vals[i], vals[i + 1]);
			i++;
		}
	}

	/*
	Checks whether the most optimal position for x is at the end,
	if it is starts from the end.

> 	*/

Reviewer: ```suggestion
	/**
	 * @return whether the most optimal position for x is
	 * at the end, if it is starts from the end.
	*/

	bool reaches_end(int x) {
		vals.push_back(x);
		int i = vals.size() - 1;
		while (i > 0 && abs(vals[i] - vals[i - 1]) <= k) {
			swap(vals[i], vals[i - 1]);
			i--;
		}
		while (i < vals.size() - 1 && vals[i] > vals[i + 1]) {
			swap(vals[i], vals[i + 1]);
			i++;
		}
		bool ans = vals.back() == x;
		vals.erase(begin(vals) + i);
		return ans;
	}
};

int main() {
	cin >> n >> k;
	int len = (int)floor(sqrt(n));
	vector<Block> sqrt{Block()};

	for (int i = 0; i < n; i++) {
		int x;
		cin >> x;
		int j = sqrt.size() - 1;

		// Finds the earliest block we can insert element x by swapping to the

> 		// left

Reviewer: can you make these two comments a bit more even
like move some words from the top line to the bottom

		while (j > 0 && (x + k >= sqrt[j].hi && x - k <= sqrt[j].lo)) { j--; }

		// Handle edge case if x doesn't cross an element greater than it in
		// block j while swapping left Swaps element x to the right to the
		// lexographically optimal position
		if (j < sqrt.size() - 1 && sqrt[j].reaches_end(x)) {
			j++;
			while (j < sqrt.size() - 1 && x >= sqrt[j].hi) { j++; }
		}
		sqrt[j].add(x);

		// If block is too large split into two smalller blocks of equal size
		if (sqrt[j].vals.size() > len) {
			int half = (sqrt[j].vals.size()) / 2;
			vector<int> left, right;
			for (int k = 0; k < half; k++) { left.push_back(sqrt[j].vals[k]); }
			for (int k = half; k < sqrt[j].vals.size(); k++) {
				right.push_back(sqrt[j].vals[k]);
			}
			sqrt.insert(begin(sqrt) + j, Block(left));
			sqrt[j + 1] = Block(right);
		}
	}

> 	for (Block block : sqrt) {

Reviewer: ```suggestion

	for (Block block : sqrt) {
```

---
id: usaco-1188
source: USACO Platinum 2022 January
title: Minimizing Haybales
author: Aakash Gokhale
---

[External Solution (C++)](http://www.usaco.org/current/data/sol_prob1_platinum_jan22.html)

## Explanation

> There are a lot of different ways to solve this problem. I am going to explain one that involves square root decomposition but I would recommend reading the official editorial to check out some of the other solutions as well.

Reviewer: ```suggestion
There's a ton of ways to solve this, but since you're probably here from the square root decomposition module, let's just go over that one.
The official editorial does have some nice alternative solutions as well.
```

> Let's first try to solve the subtask and come up with an $\mathcal{O}(N^2)$ solution.

Reviewer: maybe add some headers or something

We can insert this element into the square root block at its correct position by swapping it down to the left inside the block till it reaches a place where it can't be swapped down anymore. This also takes $\mathcal{O}(\sqrt{N})$ time because of the block size.

Now we need to swap it back towards the right until it reaches a point where which makes the array lexographically minimum. We can do this by finding the first block to the right of it that has it's $MAX > a[i]$ by checking each block again in $\mathcal{O}(\sqrt{N})$.

Notice that there is an edge case here where $a[i]$ can never be swapped down to the left past another element that is greater than it because of the restriction of $k$ even thought the $MAX$ element is greater than $a[i]$. In this case, we can push $a[i]$ into the next block and then apply the strategy of checking when the $MAX > a[i]$. We can check this by seeing that if after inserting $a[i]$ into the block and swapping towards the left and right appropriately as described above, $a[i]$ reaches the end of the block.

> Now the problem arises that when we add enough elements, the size of a block could potentially be bigger than $\sqrt{n}$. To deal with this issue we can simply split the block into two smaller blocks (each half the size of the original block) and insert them into the list of blocks. Inserting into a vector in C++ or ArrayList in Java, is dependent on the size of the array which in our case which takes $\mathcal{O}(\sqrt{N})$. We can prove that the amount of blocks is bounded to $2\sqrt{n}$.

Reviewer: please don't use "simply" or "clearly" in your editorial

Reviewer: also use `vector` and `ArrayList` instead of the raw text, looks nicer

Processing each element of the array took $\mathcal{O}(\sqrt{N})$ time so the total time complexity is $\mathcal{O}(N\sqrt{N})$.

## Implementation

**Time Complexity:** $\mathcal{O}(N\sqrt{N})$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int n, k;

struct Block {
	int hi = -INT32_MAX;
	int lo = INT32_MAX;
	vector<int> vals;

	Block(vector<int> a) {

> 		hi = -INT32_MAX;
> 		lo = INT32_MAX;

Reviewer: ```suggestion
```

---
id: usaco-1188
source: USACO Platinum 2022 January
title: Minimizing Haybales
author: Aakash Gokhale
---

[External Solution (C++)](http://www.usaco.org/current/data/sol_prob1_platinum_jan22.html)

## Explanation

There's a ton of ways to solve this, but since you're probably here from the square root decomposition module, let's just go over that one.
The official editorial does have some nice alternative solutions as well.

> ## Subtask

Reviewer: make this header & the "Full solution" one three hashtags
